core/src/main/java/org/apache/iceberg/avro/ValueReaders.java

```diff
@@ -582,13 +585,37 @@ protected StructReader(List<ValueReader<?>> readers, Types.StructType struct, Ma
         if (idToConstant.containsKey(field.fieldId())) {
           positionList.add(pos);
           constantList.add(idToConstant.get(field.fieldId()));
+        } else if (field.fieldId() == MetadataColumns.ROW_POSITION.fieldId()) {
+          // replace the _pos field reader with a position reader
+          // only if the position reader is set and this is a top-level field
```

Comment:
Not sure I understand this comment line
> only if the position reader is set

reader is set where?
> and this is a top-level field

I don't see where the top-level field restriction comes from

Reply:
This is out of date. I had to move the check because the position reader is set after this constructor. I'll update the comment.